engine/consensus/sealing/engine.go

```diff
@@ -140,123 +110,157 @@ func NewEngine(log zerolog.Logger,
-	// nil sink means nothing to send, this prevents blocking on select
-	fetchEvent := func() (*Event, EventSink, *fifoqueue.FifoQueue) {
-		if val, ok := e.pendingReceipts.Head(); ok {
-			return val.(*Event), e.receiptSink, e.pendingReceipts
-		}
-		if val, ok := e.pendingRequestedApprovals.Head(); ok {
-			return val.(*Event), e.requestedApprovalSink, e.pendingRequestedApprovals
+func (e *Engine) processAvailableMessages() error {
+
+	for {
+		// TODO prioritization
+		// eg: msg := engine.SelectNextMessage()
+		msg, ok := e.pendingRequestedApprovals.Get()
+		if !ok {
+			msg, ok = e.pendingApprovals.Get()
 		}
-		if val, ok := e.pendingApprovals.Head(); ok {
-			return val.(*Event), e.approvalSink, e.pendingApprovals
+
+		if ok {
+			e.onApproval(msg.OriginID, msg.Payload.(*flow.ResultApproval))
+			continue
 		}
-		return nil, nil, nil
+
+		// when there is no more messages in the queue, back to the loop to wait
+		// for the next incoming message to arrive.
+		return nil
 	}
+}
 
+func (e *Engine) loop() {
 	for {
-		pendingEvent, sink, fifo := fetchEvent()
 		select {
-		case event := <-e.pendingEventSink:
-			e.processPendingEvent(event)
-		case sink <- pendingEvent:
-			fifo.Pop()
-			continue
 		case <-e.unit.Quit():
 			return
+		case <-e.messageHandler.GetNotifier():
+			err := e.processAvailableMessages()
+			if err != nil {
+				e.log.Fatal().Err(err).Msg("internal error processing queued message")
+			}
 		}
 	}
 }
 
-// processPendingEvent saves pending event in corresponding queue for further processing by `Core`.
-// While this function runs in separate goroutine it shouldn't do heavy processing to maintain efficient data polling/pushing.
-func (e *Engine) processPendingEvent(event *Event) {
-	switch event.Msg.(type) {
-	case *flow.ExecutionReceipt:
-		e.engineMetrics.MessageReceived(metrics.EngineSealing, metrics.MessageExecutionReceipt)
-		e.pendingReceipts.Push(event)
-	case *flow.ResultApproval:
-		e.engineMetrics.MessageReceived(metrics.EngineSealing, metrics.MessageResultApproval)
-		if e.requiredApprovalsForSealConstruction < 1 {
-			// if we don't require approvals to construct a seal, don't even process approvals.
-			return
-		}
-		e.pendingApprovals.Push(event)
-	case *messages.ApprovalResponse:
-		e.engineMetrics.MessageReceived(metrics.EngineSealing, metrics.MessageResultApproval)
-		if e.requiredApprovalsForSealConstruction < 1 {
-			// if we don't require approvals to construct a seal, don't even process approvals.
-			return
+// processIncorporatedResult is a function that creates incorporated result and submits it for processing
+// to sealing core. In phase 2, incorporated result is incorporated at same block that is being executed.
+// This will be changed in phase 3.
+func (e *Engine) processIncorporatedResult(result *flow.ExecutionResult) {
+	e.workerPool.Submit(func() {
+		incorporatedResult := flow.NewIncorporatedResult(result.BlockID, result)
+		err := e.core.ProcessIncorporatedResult(incorporatedResult)
```

Comment:
If there are many incoming results to process, but we don't have enough workers to process them, what would happen? would the function block or buffer the call in memory and take another?

If it's the latter, then I'm a bit concerned it could cause out of memory

Reply:
I don't think it will block. Nevertheless, I find that the `WorkerPool` adds quite some (potentially unnecessary) complexity to the engine (see [my comment below](https://github.com/onflow/flow-go/pull/736#pullrequestreview-666148936) for more details)